src/retry/mod.rs

```diff
@@ -0,0 +1,235 @@
+                let elapsed_time_ms = start_time_fn.elapsed().as_millis();
+
+                if last_print_time.elapsed() >= Duration::from_secs(300) {
+                    tracing::info!("Elapsed time to retry transactions (milliseconds): {}", elapsed_time_ms);
+                    last_print_time = Instant::now();
+                }
+
+                tokio::time::sleep(Duration::from_secs(*RETRY_TX_INTERVAL)).await;
+            }
+        });
+    }
+
+    /// Retries and prunes pending transactions.
+    async fn retry_and_prune(&self) -> Result<()> {
+        let pending_transactions = self.pending_transactions().await?;
+        for transaction in pending_transactions {
+            if self.should_retry(&transaction).await? {
+                self.retry_transaction(transaction).await?;
+            } else {
+                self.prune_transaction(transaction.tx.hash).await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Retries a transaction and prunes it if the conversion to a primitive transaction fails.
+    async fn retry_transaction(&self, transaction: StoredPendingTransaction) -> Result<()> {
+        let hash = transaction.tx.hash;
+        tracing::info!("Retrying transaction {hash} with {} retries", transaction.retries + 1);
+
+        // Generate primitive transaction, handle error if any
+        let transaction = match TransactionSignedEcRecovered::try_from(transaction.tx.clone()) {
+            Ok(transaction) => transaction,
+            Err(error) => {
+                self.prune_transaction(hash).await?;
+                return Err(error.into());
+            }
+        };
+
+        let _hash = self.eth_provider.send_raw_transaction(transaction.into_signed().envelope_encoded()).await?;
+        #[cfg(test)]
+        self.retried.lock().await.push(_hash);
+        Ok(())
+    }
+
+    /// Prunes a transaction from the database given its hash.
+    async fn prune_transaction(&self, hash: B256) -> Result<()> {
+        tracing::info!("Pruning pending transaction: {hash}");
+        let filter = EthDatabaseFilterBuilder::<filter::Transaction>::default().with_tx_hash(&hash).build();
+        self.database.delete_one::<StoredPendingTransaction>(filter).await?;
+        Ok(())
+    }
+
+    /// Returns all pending transactions.
+    async fn pending_transactions(&self) -> Result<Vec<StoredPendingTransaction>> {
+        Ok(self.database.get(None, None).await?)
+    }
+
+    /// Returns true if the transaction should be retried. A transaction should be retried if it has
+    /// not been executed and the number of retries is less than the maximum number of retries.
+    async fn should_retry(&self, transaction: &StoredPendingTransaction) -> Result<bool> {
+        let max_retries_reached = transaction.retries + 1 >= *TRANSACTION_MAX_RETRIES;
+        let transaction_executed = self.database.transaction(&transaction.tx.hash).await?.is_some();
+        Ok(!max_retries_reached && !transaction_executed)
```

Comment:
Here perhaps we should make the assertions in the other direction to avoid `!` (which are less idiomatic) but perhaps it would also be less clear, what do you think?

Reply:
I wrote it in the way that was the clearer for me, I think this makes it the more readable no? I can change if you prefer

Reply:
No this is fine let us keep like this